[RFC] documentation for manual struct edits

TheOneric opened the pull request:
There was previously no documentation at all and to me no clear assumption which could be derived from the code. So this will define the API itself and not just be documentation of existing assumptions. I'd appreaciate if everyone could give some feedback on the restrictions layed out by these patches.

Also as a side-product, I noticed that `treat_fontname_as_pattern`/`treat_family_as_pattern` which was introduced in 256c11324911da42ff829d9cb6906186b5173464 no longer seems to do anything, though a few vestiges remain in ass_render.c and the (very outdated) documentation of `ass_font_select`. Before these emains are removed completly, can someone verify it is actually doing nothing and supposed to be like this?

rcombs commented on libass/ass_types.h:
```diff
@@ -98,32 +98,32 @@ typedef struct ass_library ASS_Library;
-    int Justify;
+    int treat_fontname_as_pattern; // does nothing (left in place for ABI-compatibility)
+    double Blur; // sets a default \blur for the event, same values as \blur
+    int Justify; // sets text justification independent of event alignment, use ASS_JUSTIFY_*
```
Semicolons rather than commas here

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ * The main use case for this is _reading_ the track fields, especially
+ * track->YCbCrMatrix, to correctly display the rendering results.
+ *
+ * Furthermore the exposed definitions also open up the possibility to _modify_
```
Would be worth sprucing up the grammar here a bit: "Furthermore", "e.g.", "well-versed", "internals,", "API breaks"…

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ * workflows that make use of this possibility are also more likely to be
+ * affected by future API-breaks than those which do not.
+ *
+ * To avoid desynchronisation with internal states there are some restrictions
```
"states,",

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ * Ignoring them may lead to undefined behaviour. See the following listing:
+ *
+ *  - Manual struct edits and track-modifying (including modification to the
+ *    event and style elements of the track) API-calls cannot be freely mixed:
```
"API calls"

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *  - Manual struct edits and track-modifying (including modification to the
+ *    event and style elements of the track) API-calls cannot be freely mixed:
+ *    - Before manual changes are performed it is allowed to call any such API,
+ *      unless the documentation of the API-call says otherwise.
```
"function"

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *
+ *  - Manual struct edits and track-modifying (including modification to the
+ *    event and style elements of the track) API-calls cannot be freely mixed:
+ *    - Before manual changes are performed it is allowed to call any such API,
```
"performed,"

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *    event and style elements of the track) API-calls cannot be freely mixed:
+ *    - Before manual changes are performed it is allowed to call any such API,
+ *      unless the documentation of the API-call says otherwise.
+ *    - After manual changes were performed, no track-modifying API must be
```
"changes have been performed"; "API may be"

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *    - Before manual changes are performed it is allowed to call any such API,
+ *      unless the documentation of the API-call says otherwise.
+ *    - After manual changes were performed, no track-modifying API must be
+ *      invoked any more, except for ass_track_set_feature and ass_flush_events.
```
drop "any more"

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *      unless the documentation of the API-call says otherwise.
+ *    - After manual changes were performed, no track-modifying API must be
+ *      invoked any more, except for ass_track_set_feature and ass_flush_events.
+ *  - After rendering started existing array members, eg members of events, and
```
"After the first call to ass_render_frame, existing array members (e.g. members of events) and non-array track fields (e.g. PlayResX or event_format) must not be modified."

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *    all non-array track fields, like eg PlayResX or event_format, must not
+ *    be modified any more.
+ *    Adding _new_ members to arrays and updating
+ *    the corresponding counter remains allowed.
```
Wrap?

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *    Adding _new_ members to arrays and updating
+ *    the corresponding counter remains allowed.
+ *  - Adding and removing members to array fields, like events or styles,
+ *    must be done through the corresponding API-call, eg ass_alloc_event.
```
"API function, e.g."

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *    the corresponding counter remains allowed.
+ *  - Adding and removing members to array fields, like events or styles,
+ *    must be done through the corresponding API-call, eg ass_alloc_event.
+ *    Also see the documentation of these API-calls.
```
"See also the documentation of these functions."

rcombs commented on libass/ass_types.h:
```diff
@@ -22,6 +22,52 @@
+ *  ass_process_chunk, ass_read_styles, ...
+ *
+ * Direct struct modification can be done safely, but it is also easy to
+ * miss an initialisation or violate the restrictions introducing bugs
```
"violate these restrictions, thus introducing bugs…"

rcombs commented on libass/ass.h:
```diff
@@ -597,6 +599,10 @@ int ass_alloc_event(ASS_Track *track);
  * \param sid style id
  * Deallocates style data. Does not modify track->n_styles.
+ * Freeing a style without subsequently setting track->n_styles
+ * to a value smaller or equal to the freed style id before
```
"less than or equal to"

rcombs commented on libass/ass.h:
```diff
@@ -605,6 +611,10 @@ void ass_free_style(ASS_Track *track, int sid);
  * Deallocates event data. Does not modify track->n_events.
+ * Freeing an event without subsequently setting track->n_events
+ * to a value smaller or equal to the freed event id before rendering
+ * the next frame or calling ass_flush_events is undefined behaviour.
```
just say calling any other libass API function on the track

rcombs commented on libass/ass.h:
```diff
@@ -564,6 +564,7 @@ ASS_Track *ass_new_track(ASS_Library *);
  * headers. (ass_process_chunk() will not change any of the flags.)
  * Additions to ASS_Feature are backward compatible to old libass releases (ABI
  * compatibility).
+ * Calling this after rendering of the track started is disallowed.
```
"after calling ass_render_frame"?

rcombs commented on libass/ass_types.h:
```diff
@@ -98,32 +98,32 @@ typedef struct ass_library ASS_Library;
-    char *Name;
-    char *FontName;
+    char *Name;     //must be a valid non-NULL string pointer
+    char *FontName; //must be a valid non-NULL string pointer
```
Specify how these should be allocated? afaik they have to come from the same copy of `malloc` as libass uses (which has some implications on windows)

Also, worth noting that while they can't be null, they can point to empty strings.

rcombs commented on libass/ass_types.h:
```diff
@@ -242,9 +242,9 @@ typedef struct ass_track {
-    char *Language;
+    int ScaledBorderAndShadow; // 0 or 1 (boolean)
+    int Kerning; // 0 or 1 (boolean)
+    char *Language; // zero-terminated ISO-639-1 codes
```
this can be null, right? also, singular "code"

TheOneric replied:
thanks, removed the plural. Yes, NULL is allowed here.

rcombs reviewed (changes requested):
Just bikeshedding on phrasing and such; conceptually I like the clarifications.

CastagnaIT commented on libass/ass_types.h:
```diff
@@ -52,32 +99,32 @@ typedef struct ass_library ASS_Library;
+    double ScaleY; // positive with 1.0 representing 100%
     double Spacing;
     double Angle;
     int BorderStyle;
```
it would be useful to document BorderStyle as well, and i could suggest to add a borderstyle enum values or something like constants to be used to set this property

TheOneric replied:
The `BorderStyle` values are identical to ASS + libass' ASS extensions. I don't think we need to repeat ASS(-extension) spec/behaviour in each field's doc-comment.

TheOneric commented:
Added
```
 * Additionally if styles a re freed such that an for an event ev in track->events
 * ev->Style >= track->n_styles will also result in undefined behaviour.
```
to the documentation of `ass_free_style`.

TheOneric commented:
Rebased, added changes to docs of `ass_render_frame`'s `detect_change` parameter and applied the “api: do not leak deprecation macros”-commit to master to hopefully avoid anyone starting to use those.

TheOneric commented:
If anyone is planning to take (another) look at this, say so now and I'll wait for whatever amount of time is required. Otherwise I'll assume implicit approval and merge this in ~2 days.
